src/charm.py

```diff
@@ -12,92 +12,292 @@
+        env = Environment(loader=FileSystemLoader(self.template_dir))
+
+        resources = [
+            env.get_template(template).render(self._stored.ceph_data)
+            for template in self.RESOURCE_TEMPLATES
+        ]
+
+        resource_dicts = [yaml.safe_load_all(res) for res in resources]
+        return list(itertools.chain.from_iterable(resource_dicts))
+
+    def render_storage_definitions(self) -> List[Dict]:
+        """Render StorageClass definitions for supported filesystem types."""
+        env = Environment(loader=FileSystemLoader(self.template_dir))
+        default_storage = self._stored.default_storage_class
+        storage_classes = []
+
+        ext4_ctx = self.copy_stored_dict(self._stored.ceph_data)
+        ext4_ctx["default"] = default_storage == self.EXT4_STORAGE
+        ext4_ctx["pool_name"] = "ext4-pool"
+        ext4_ctx["fs_type"] = "ext4"
+        ext4_ctx["sc_name"] = "ceph-ext4"
+        resource = env.get_template(self.STORAGE_CLASS_TEMPLATE).render(ext4_ctx)
+        storage_classes.append(yaml.safe_load(resource))
+
+        xfs_ctx = self.copy_stored_dict(self._stored.ceph_data)
+        xfs_ctx["default"] = default_storage == self.XFS_STORAGE
+        xfs_ctx["pool_name"] = "xfs-pool"
+        xfs_ctx["fs_type"] = "xfs"
+        xfs_ctx["sc_name"] = "ceph-xfs"
+        resource = env.get_template(self.STORAGE_CLASS_TEMPLATE).render(xfs_ctx)
+        storage_classes.append(yaml.safe_load(resource))
+
+        return storage_classes
+
+    def render_all_resource_definitions(self) -> List[Dict]:
+        """Render all resources required for ceph-csi."""
+        return self.render_resource_definitions() + self.render_storage_definitions()
+
+    @needs_leader
+    def update_storage_classes(self) -> None:
+        """Re-render templates and update StorageClass resources in k8s cluster."""
+        for resource in self.resources:
+            if isinstance(resource, StorageClass):
+                resource.remove()
+        storage_classes = self.render_storage_definitions()
+        self.create_ceph_resources(storage_classes)
+
+    @needs_leader
+    def _on_ceph_joined(self, event: RelationJoinedEvent) -> None:
+        """Create necessary k8s resources when relation is formed with ceph-mon."""
+        if self._stored.resources_created:
+            # Skip silently if other ceph_relation_joined event already
+            # created resources
+            return
+
+        unit_data = event.relation.data[event.unit]
+        expected_data = ("fsid", "key", "mon_hosts")
+        for key in expected_data:
+            self._stored.ceph_data[key] = unit_data.get(key)
+        missing_data = [key for key, value in self._stored.ceph_data.items() if value is None]
+        if missing_data:
+            logger.warning(
+                "Ceph relation with %s is missing data: %s", event.unit.name, missing_data
+            )
+            self.unit.status = BlockedStatus("Ceph relation is missing data.")
```

Comment:
Should this be `WaitingStatus`? If a user needs to run any step against this unit, then `blocked` is accurate. If the user needs to wait for other processes within Juju to be completed (e.g. queue hooks that need to be triggered in this unit), then `waiting` should be the status.

Could this happen because the Ceph units are being bootstrapped / configured, so they're not ready to serve other applications?

Reply:
Good point, this should be `WaitingStatus`. I also opened #4 because we need a way to recover from this `Waiting` state.